google/datalab/contrib/mltoolbox/_local_predict.py

```diff
@@ -158,11 +165,99 @@ def get_prediction_results(model_dir, data, headers, img_cols=None, show_image=T
+
+    predict_results = ml.ModelVersions(parts[0]).predict(parts[1], predict_data)
+  else:
+    predict_results = _tf_predict(model_dir_or_id, predict_data)
+
   df_r = pd.DataFrame(predict_results)
   df_s = pd.DataFrame(display_data)
   df = pd.concat([df_r, df_s], axis=1)
   # Remove duplicate columns. All 'key' columns are duplicate here.
   df = df.loc[:, ~df.columns.duplicated()]
 
   return df
+
+
+def _batch_csv_reader(csv_file, n):
+  with file_io.FileIO(csv_file, 'r') as f:
+    args = [f] * n
+    return itertools.izip_longest(*args)
+
+
+def _get_output_schema(session, output_alias_map):
+  schema = []
+  for name in sorted(six.iterkeys(output_alias_map)):
+    tensor_name = output_alias_map[name]
+    dtype = session.graph.get_tensor_by_name(tensor_name).dtype
+    if dtype == tf.int32 or dtype == tf.int64:
+      schema.append({'name': name, 'type': 'INTEGER'})
+    elif dtype == tf.float32 or dtype == tf.float64:
+      schema.append({'name': name, 'type': 'FLOAT'})
+    else:
+      schema.append({'name': name, 'type': 'STRING'})
+
+  return schema
+
+
+def _format_results(output_format, output_schema, batched_results):
+  # Convert a dict of list to a list of dict.
+  batched_results = [dict(zip(batched_results, t)) for t in zip(*batched_results.values())]
+  if output_format == 'csv':
+    results = []
+    for r in batched_results:
+      values = [str(r[schema['name']]) for schema in output_schema]
+      results.append(','.join(values))
+  elif output_format == 'json':
+    # Default json encoder cannot handle numpy types.
+    class _JSONEncoder(json.JSONEncoder):
+
+      def default(self, obj):
+        if isinstance(obj, np.integer):
+            return int(obj)
+        elif isinstance(obj, np.floating):
+            return float(obj)
+        elif isinstance(obj, np.ndarray):
+            return obj.tolist()
+        else:
+            return super(_JSONEncoder, self).default(obj)
+
+    results = [json.dumps(x, cls=_JSONEncoder) for x in batched_results]
+  else:
+    raise ValueError('Unknown output_format %s' % output_format)
+
+  return results
+
+
+def local_batch_predict(model_dir, csv_file_pattern, output_file, output_format):
+  """ Batch Predict with a specified model.
+
+  It does batch prediction, saves results to output_file and also creates an output
+  schema file.
+
+  Args:
+    model_dir: The model directory containing a SavedModel (usually saved_model.pb).
+    csv_file_pattern: a pattern of csv files as batch prediction source.
+    output_file: the path of file to output.
+    output_format: csv or json.
+  """
+  csv_files = file_io.get_matching_files(csv_file_pattern)
+  with tf.Graph().as_default(), tf.Session() as sess:
+    input_alias_map, output_alias_map = _tf_load_model(sess, model_dir)
+    csv_tensor_name = list(input_alias_map.values())[0]
+    output_schema = _get_output_schema(sess, output_alias_map)
+    with file_io.FileIO(output_file, 'w') as f:
```

Comment:
should the output be sharded if the input is? What if 100 1GB csv files are passed in? Maybe the number of shards could equal the number of input shards